File: frontend/src/api/access.ts
import { apiFetch } from 'src/api/client';

/**
 * One user's roles and row-level scope — mirrors `AccessOut` in
 * backend/app/api/schemas.py. SEMANTIC_LAYER.md § 19: `scopeStates` confines
 * every clinical query this user makes (chat, saved questions, everything) to
 * patients in those states; `null` is unconfined. `availableStates` comes
 * from the loaded dataset rather than a hardcoded list of fifty.
 */
export interface Access {
  userId: string;
  roles: string[];
  scopeStates: null | string[];
  availableStates: string[];
}

export const accessKeys = {
  me: ['access', 'me'] as const,
};

export const fetchMyAccess = () => apiFetch<Access>('/access/me');

export const updateMyAccess = (scopeStates: null | string[]) =>
  apiFetch<Access>('/access/me', { body: JSON.stringify({ scopeStates }), method: 'PUT' });

File: backend/app/api/routes/access.py
"""Row-level access, in its smallest honest form — SEMANTIC_LAYER.md § 19.

Self-service only: `PUT` changes the caller's own `scope_states`, never
someone else's — there is no user id in the body, so "edit someone else's
scope" is not even expressible here. With one hardcoded dev user this is a
way to prove the seam end to end (pick a state, watch every subsequent cohort
confine to it) rather than a permission model; see
`app/services/authz_service.py`'s own docstring on that distinction.

The write is gated `RequireCurator`, the same guard a definition edit gets:
changing what a user's own queries can see is exactly the kind of change
CONVENTIONS.md's authorization discipline exists to keep deliberate.
"""

from fastapi import APIRouter

from app.api.deps import CurrentUser, DbSession, RequireCurator
from app.api.schemas import AccessOut, AccessUpdateIn, ApiResponse
from app.services import authz_service, catalog_service

router = APIRouter(prefix="/access", tags=["access"])


async def _access_out(session: DbSession, *, user_id: str) -> AccessOut:
    roles = await authz_service.get_roles(session, user_id=user_id)
    scope_states = await authz_service.get_scope_states(session, user_id=user_id)
    available_states = await catalog_service.list_patient_states(session)
    return AccessOut(
        available_states=list(available_states),
        roles=sorted(roles),
        scope_states=scope_states,
        user_id=user_id,
    )


@router.get("/me")
async def get_my_access(session: DbSession, user_id: CurrentUser) -> ApiResponse[AccessOut]:
    return ApiResponse(data=await _access_out(session, user_id=user_id))


@router.put("/me")
async def update_my_access(
    body: AccessUpdateIn, session: DbSession, user_id: RequireCurator
) -> ApiResponse[AccessOut]:
    # Roles are untouched by this route (see AccessUpdateIn) — carry them
    # forward exactly as they are rather than requiring the caller to repeat
    # them back, which is how a client that forgot one silently drops a role.
    roles = await authz_service.get_roles(session, user_id=user_id)
    await authz_service.set_roles(
        session, user_id=user_id, roles=sorted(roles), scope_states=body.scope_states
    )
    return ApiResponse(data=await _access_out(session, user_id=user_id))
